core/src/main/java/org/apache/calcite/sql/validate/implicit/TypeCoercionImpl.java

```diff
@@ -563,6 +564,76 @@ protected boolean booleanEquality(SqlCallBinding binding,
     return false;
   }
 
+  /**
+   * {@inheritDoc}
+   *
+   * <p>STRATEGIES
+   *
+   * <p>To determine the common type:
+   *
+   * <ul>
+   *
+   * <li>When the LHS has a Simple type and RHS has a Collection type determined by {@link SqlTypeUtil#isCollection},
+   * to find the common type of LHS's type and RHS's component type.
+   * <li>If the common type differs from the LHS type, then coerced LHS type,
+   * and nullable remains unchanged.
+   * <li>Create a new Collection type that matches the common type,
+   * and nullable keep same as RHS's component type and RHS's collection type.
+   * <li>If this new Collection type differs from the RHS type, adjust the RHS type as needed.
+   *
+   *<pre>
+   * field1       ARRAY(field2, field3, field4)
+   *    |                |       |        |
+   *    |               +-------+---------+
+   *    |                       |
+   *    |                 component type
+   *    |                      |
+   *    +-----common type------+
+   *</pre>
+   *
+   * <li>Notice: If either LHS or RHS has a {@link SqlTypeName#NULL} type, it will directly return without any adjusting.
+   * </ul>
+   */
+  @Override public boolean quantifyOperationCoercion(SqlCallBinding binding) {
+    final RelDataType type1 = binding.getOperandType(0);
+    final RelDataType collectionType = binding.getOperandType(1);
+    final RelDataType type2 = collectionType.getComponentType();
+    assert type2 != null;
+    if (type1.getSqlTypeName() == SqlTypeName.NULL || type2.getSqlTypeName() == SqlTypeName.NULL) {
+      return false;
+    }
+    final SqlNode node1 = binding.operand(0);
+    final SqlNode node2 = binding.operand(1);
+    RelDataType widenType = commonTypeForBinaryComparison(type1, type2);
+    if (widenType == null) {
+      widenType = getTightestCommonType(type1, type2);
+    }
+    if (widenType == null) {
+      return false;
+    }
+    final RelDataType leftWidenType =
+        binding.getTypeFactory().enforceTypeWithNullability(widenType, type1.isNullable());
+    boolean coercedLeft =
+        coerceOperandType(getScope(binding), binding.getCall(), 0, leftWidenType);
+    if (coercedLeft) {
+      updateInferredType(node1, leftWidenType);
+    }
+    final RelDataType rightWidenType =
+        binding.getTypeFactory().enforceTypeWithNullability(widenType, type2.isNullable());
+    RelDataType collectionWidenType =
+        binding.getTypeFactory().createArrayType(rightWidenType, -1);
+    collectionWidenType =
+        binding
+            .getTypeFactory()
+            .enforceTypeWithNullability(collectionWidenType, collectionType.isNullable());
+    boolean coercedRight =
+        coerceOperandType(getScope(binding), binding.getCall(), 1, collectionWidenType);
```

Comment:
I hope these calls adjust the types in the typeMap and the callbinding too.

Reply:
Are you saying here that we don't need to update the type of "call" again, instead update it directly within "coerceOperandType"?

Reply:
I understand what you mean. We can discuss this in creating a new PR It's not a problem that can be quickly modified.

Reply:
I want to avoid the bug as in https://issues.apache.org/jira/browse/CALCITE-6743

Reply:
The RHS type has been updated by:
```
if (coercedRight) {
      updateInferredType(node2, collectionWidenType);
    } 
```
I'm not sure if this is the same issue, but it would be helpful to add a description of the usage issues that may arise when having inaccurate types in CALCITE-6723.
 I will also review another PR mentioned in the comments on CALCITE-6723, as there may be some duplicate code that can achieve the same capability. Currently, I am unsure.

Reply:
After 6743 the program includes casts, but the validator types in the typeMap are still the old ones, before the casts were inserted.

Reply:
I will check this.

Reply:
This SQL:
`select 1.0 = some (ARRAY[2,3,null]);`
The nodeToTypeMap is:
![Screen Shot 2025-05-11 at 08 45 12](https://github.com/user-attachments/assets/0ccc1515-5e65-44b6-aca5-ff172580835b)

Our question is Which of the following three are legal？

1: ARRAY [CAST (2 AS INTEGER), CAST (3 AS INTEGER), NULL] ->DECIMAL (11,1) ARRAY
2: CAST(ARRAY[CAST(2 AS INTEGER), CAST(3 AS INTEGER), NULL] AS DECIMAL(11, 1) ARRAY) -> DECIMAL (11,1) ARRAY
3: ARRAY [CAST(CAST (2 AS INTEGER) as DECIMAL(11,1)), CAST(CAST (3 AS INTEGER) as  DECIMAL (11,1)), CAST(NULL as DECIMAL(11,1))]->DECIMAL (11,1) ARRAY

Reply:
which one is "legal" depends on the dialect that the function belongs to; the existing behavior attempted to model Spark.
But the main requirement is for the nodes that end up in the Rel tree to all be present in the nodeToTypeMap.

Reply:
According to the screenshot above, does the current PR also have the same issue? The node `ARRAY [CAST (2 AS INTEGER), CAST (3 AS INTEGER), NULL] ->DECIMAL (11,1) ARRAY` has been updated.

Reply:
@mihaibudiu Does this discussion affect the merging of the current issues?

Reply:
I was assuming this pr is fine and that you have checked

Reply:
According to my understanding, I have checked it.